### PR TITLE
add signature graph for parsing an identities key history

### PR DESCRIPTION
**Pull request by purehyperbole**

### CODE REVIEW

File: lib/signature_graph.rb [hunk 1]
```diff
@@ -0,0 +1,263 @@
...
+      @previous = op[:previous]
+      @timestamp = op[:timestamp]
+      @version = op[:version]
+      @actions = op[:actions]
+      @signing_key = hdr[:kid]
+
+      validate
+    end
+
+    def validate
```
Comment (adriacidre) [inline lib/signature_graph.rb:29]: In general if the method throws an exception you name it with a bang at the end like `validate!` so the consumer can prepare its code to catch the exception.

File: Gemfile.lock [hunk 2]
```diff
@@ -35,7 +35,7 @@ GEM
     faye-websocket (0.11.0)
       eventmachine (>= 0.12.0)
       websocket-driver (>= 0.5.1)
-    google-protobuf (3.13.0-universal-darwin)
```
Comment (adriacidre) [inline Gemfile.lock:38]: I'm not sure if there is a way to add a rule to ignore this stuff.

File: lib/signature_graph.rb [hunk 3]
```diff
@@ -0,0 +1,271 @@
...
+      raise "operation does not specify a previous signature" if @previous.nil?
+      raise "invalid operation timestamp" if @timestamp < 1
+      raise "operation does not specify any actions" if @actions.nil?
+      raise "operation does not specify any actions" if @actions.length < 1
+      raise "operation does not specify an identifier for the signing key" if @signing_key.nil?
+    end
+
+    def revokes(kid)
+      @actions.each do |action|
+        return true if action[:kid] == kid && action[:action] = "key.revoke"
```
Comment (adriacidre) [inline lib/signature_graph.rb:43]: I think the second operand in this conditional has a missing `=`, should be `==` instead.

As the conditional is not a single condition may be better to write it like
```
if a=="a" && b=="b"
  return true
end
```
that will make it easy to understand

File: lib/signature_graph.rb [hunk 4]
```diff
@@ -0,0 +1,271 @@
...
+      k
+    end
+
+    def key_by_device(did)
+      k = @devices[did]
+      raise "key not found" if k.nil?
+      k
+    end
+
+    def execute(operation)
```
Comment (adriacidre) [inline lib/signature_graph.rb:116]: would be nice to refactor / split this method

File: lib/signature_graph.rb [hunk 5]
```diff
@@ -0,0 +1,271 @@
...
+
+        if action[:type] != "device.key" && action[:type] != "recovery.key"
+          raise "operation action does not provide a valid type"
+        end
+
+        if action[:action] != "key.add" && action[:action] != "key.revoke"
+          raise "operation action does not provide a valid action"
+        end
+
+        if action[:action] == "key.add" && action[:key].nil?
```
Comment (adriacidre) [inline lib/signature_graph.rb:154]: what do you think about moving `key.add` `key.revoke`...  to constants?

File: lib/signature_graph.rb [hunk 6]
```diff
@@ -0,0 +1,271 @@
...
+          raise "operation action does not provide a valid public key"
+        end
+
+        if action[:action] == "key.add" && action[:type] == "device.key" && action[:did].nil?
+          raise "operation action does not provide a valid device id"
+        end
+
+        raise "operation action does not provide a valid timestamp for the action to take effect from" if action[:from] < 0
```
Comment (adriacidre) [inline lib/signature_graph.rb:162]: can you move this conditional to an extended format, I had to read this twice to see the if at the end :P

File: lib/signature_graph.rb [hunk 7]
```diff
@@ -0,0 +1,271 @@
...
+      if @keys[action[:kid]].nil? != true
+        raise "operation contains a key with a duplicate identifier" 
+      end
+
+      k = Key.new(action)
+
+      case action[:type]
+      when "device.key"
+        dk = @devices[action[:did]]
+        if dk.nil? != true
```
Comment (adriacidre) [inline lib/signature_graph.rb:212]: Use `if not dk.nil?` or `unless dk.nil?` instead

File: lib/signature_graph.rb [hunk 8]
```diff
@@ -0,0 +1,271 @@
...
+          raise "operation contains more than one active key for a device" unless dk.revoked?
+        end
+      when "recovery.key"
+        if @recovery_key.nil? != true
```
Comment (adriacidre) [inline lib/signature_graph.rb:216]: Use `if not @recovery_key.nil?` or `unless @recovery_key.nil?` instead

**Review by adriacidre — COMMENTED** [review]
Left some comments, in general I would suggest split those big methods on something easy to understand. But overall looks good. Well done

File: Gemfile.lock (hunk 2 above)
Comment (purehyperbole) [reply Gemfile.lock:38]: yeah, not sure. I guess it keeps flipping depending on what os was used to last call `bundle install` ?

Comment (adriacidre) [reply Gemfile.lock:38]: yep I think so, not sure if there is a proper way to handle this stuff

**Review by adriacidre — APPROVED** [review]

**Pull request merged**